controllers/controllers.go

```diff
@@ -155,9 +155,35 @@ func (r *instanceReconciler) deconfigureInstance(node *core.Node) error {
 	return nil
 }
 
-// windowsNodePredicate returns a predicate which filters out all node objects that are not Windows nodes.
+// windowsNodeUpgradePredicate returns a predicate whose filter catches only Windows node objects that signal upgrade
+// to the latest WMCO version through a change in the node version annotation
+func windowsNodeUpgradePredicate() predicate.Funcs {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			// catch Machine-backed Windows node upgrades as they are re-created
+			return e.Object.GetLabels()[core.LabelOSStable] == "windows" &&
+				e.Object.GetAnnotations()[metadata.VersionAnnotation] == version.Get()
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			// catch BYOH Windows node upgrades to the current WMCO version as they are re-configured in place
+			return e.ObjectNew.GetLabels()[core.LabelOSStable] == "windows" &&
+				e.ObjectNew.GetAnnotations()[metadata.VersionAnnotation] == version.Get()
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return e.Object.GetLabels()[core.LabelOSStable] == "windows" &&
+				e.Object.GetAnnotations()[metadata.VersionAnnotation] == version.Get()
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			// catch if someone deletes an node stuck at an older WMCO version for example
+			return e.Object.GetLabels()[core.LabelOSStable] == "windows"
+		},
+	}
+}
+
+// outdatedWindowsNodePredicate returns a predicate which filters out all node objects that are not up-to-date Windows
+// nodes. Up-to-date refers to the version annotation and public key hash annotations.
 // If BYOH is true, only BYOH nodes will be allowed through, else no BYOH nodes will be allowed.
-func windowsNodePredicate(byoh bool) predicate.Funcs {
+func outdatedWindowsNodePredicate(byoh bool) predicate.Funcs {
```

Comment:
Did you considered extending  [isValidWindowsNode()](https://github.com/openshift/windows-machine-config-operator/blob/61e24e9f8fcb31a451ea1b2408c7382821f4a78f/controllers/controllers.go#L235)   instead?

Reply:
I don't follow, isValidWindowsNode is a helper to this method

Reply:
Gotcha. We cannot extend `isValidWindowsNode()` as it's purpose is independent of the version annotation. See how it is used in this method a few line below (`VersionAnnotation != version.Get()` whereas the newly introduced predicate is looking for the inverse case, `VersionAnnotation == version.Get()`)